File: 1481-least-number-of-unique-integers-after-k-removals/1481-least-number-of-unique-integers-after-k-removals.js
/**
 * @param {number[]} arr
 * @param {number} k
 * @return {number}
 */

// keep track of the frequency
// greedy method: keep removing from the least frequency first
// return the number of keys that don't have a value of 0
var findLeastNumOfUniqueInts = function(arr, k) {
    const freq = {}
    
    for (const num of arr) {
        freq[num] = freq[num] + 1 || 1
    }
    
    const val = Object.values(freq).sort((a, b) => a - b)
    let deletes = 0
    let i = 0
    
    while (k > 0) {
        if (val[i] <= k) {
            deletes++
            k -= val[i]
        } else {
            break;
        }
        i++
    }
    
    return val.length - deletes
};

File: 1481-least-number-of-unique-integers-after-k-removals/1481-least-number-of-unique-integers-after-k-removals.py
from collections import Counter

# get the frequency of each number
# sort the frequency in ascending order
# greedy: delete the numbers with the lowest frequency first
class Solution:
    def findLeastNumOfUniqueInts(self, arr: List[int], k: int) -> int:
        count = Counter(arr)
        sorted_count = sorted(count.values())
        
        deletions = 0
        i = 0
        
        while k > 0:
            if sorted_count[i] <= k:
                deletions += 1
                k -= sorted_count[i]
            else:
                break    
            i += 1
            
        return len(sorted_count) - deletions